File: indeed-scraper/browser/python/indeed.py
"""Indeed scraper using the official Scrapeless Python SDK + Playwright over CDP.
Indeed exposes two embedded JSON blobs:
- Search pages: `window.mosaic.providerData["mosaic-provider-jobcards"]={...}`
- Job pages:    `_initialData={...};`

Both are extracted with the same regexes the upstream reference uses and the dicts are
forwarded with identical top-level keys (`results`, `meta`, `description`, plus
the flattened `jobMetadataHeaderModel` / `jobTagModel` / `jobInfoHeaderModel`).
"""

from __future__ import annotations

import asyncio
import json
import os
import re
from typing import Any
from urllib.parse import parse_qsl, urlencode, urlparse, urlunparse

from loguru import logger
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "US"
DEFAULT_SESSION_TTL = 180

SEARCH_JSON_RE = re.compile(
    r'window\.mosaic\.providerData\["mosaic-provider-jobcards"\]=(\{.+?\});',
    re.DOTALL,
)
JOB_JSON_RE = re.compile(r"_initialData=(\{.+?\});", re.DOTALL)

# ---------------------------------------------------------------------------
# Scrapeless plumbing
# ---------------------------------------------------------------------------

def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()

def _looks_auth_walled(html: str, final_url: str) -> bool:
    if not html:
        return True
    if "secure.indeed.com/auth" in (final_url or ""):
        return True
    if "Sign In | Indeed Accounts" in html:
        return True
    if "Just a moment" in html and len(html) < 90000:
        return True
    return False


async def _fetch_rendered_html(
    url: str,
    ready_selector: str | None = None,
    *,
    proxy_country: str = DEFAULT_PROXY_COUNTRY,
    retries: int = 2,
    warmup: bool = True,
) -> str:
    last_error: Exception | None = None
    for attempt in range(retries + 1):
        client = _client()
        session = client.browser.create(
            ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
        )
        ws = session.browser_ws_endpoint
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(ws)
            try:
                page = await browser.new_page()
                if warmup:
                    try:
                        await page.goto("https://www.indeed.com/", wait_until="domcontentloaded", timeout=30000)
                        for _ in range(6):
                            await asyncio.sleep(2)
                            h = await page.content()
                            if h and "Just a moment" not in h and len(h) > 30000:
                                break
                    except Exception:
                        pass
                await page.goto(url, wait_until="domcontentloaded", timeout=45000)
                html = ""
                for _ in range(6):
                    if ready_selector:
                        try:
                            await page.wait_for_selector(ready_selector, timeout=5000)
                            break
                        except Exception:
                            pass
                    await asyncio.sleep(2)
                    html = await page.content()
                    if html and "Just a moment" not in html:
                        break
                final_url = page.url
                if not html:
                    html = await page.content()
                if html and not _looks_auth_walled(html, final_url):
                    return html
                last_error = RuntimeError(
                    f"auth-wall (final={final_url[:60]})" if _looks_auth_walled(html, final_url) else "empty HTML"
                )
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("attempt {} failed: {}", attempt + 1, e)
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
    raise RuntimeError(f"giving up after {retries + 1} attempts: {last_error}")

def _add_url_parameter(url: str, **params: Any) -> str:
    parsed = urlparse(url)
    query = dict(parse_qsl(parsed.query))
    query.update({k: str(v) for k, v in params.items()})
    return urlunparse(parsed._replace(query=urlencode(query)))

# ---------------------------------------------------------------------------
# Parsers — verbatim shape from the upstream reference's indeed.py
# ---------------------------------------------------------------------------

def parse_search_page(html: str) -> dict[str, Any]:
    m = SEARCH_JSON_RE.search(html)
    if not m:
        return {"results": [], "meta": []}
    data = json.loads(m.group(1))
    model = data.get("metaData", {}).get("mosaicProviderJobCardsModel", {})
    return {
        "results": model.get("results", []),
        "meta": model.get("tierSummaries", []),
    }

def parse_job_page(html: str) -> dict[str, Any]:
    m = JOB_JSON_RE.search(html)
    if not m:
        return {}
    data = json.loads(m.group(1))
    job_data: dict[str, Any] = {}
    try:
        job_data["description"] = data["jobInfoWrapperModel"]["jobInfoModel"][
            "sanitizedJobDescription"
        ]
    except (KeyError, TypeError):
        job_data["description"] = ""
    for sub_key in ("jobMetadataHeaderModel", "jobTagModel", "jobInfoHeaderModel"):
        sub = data.get(sub_key)
        if isinstance(sub, dict):
            job_data.update(sub)
    return job_data

# ---------------------------------------------------------------------------
# Scrape functions — mirror the upstream reference's exports verbatim
# ---------------------------------------------------------------------------

async def scrape_search(url: str, max_results: int = 1000) -> list[dict[str, Any]]:
    """Scrape Indeed search results across pagination (10 jobs per page)."""
    first = await _fetch_rendered_html(url, ready_selector="#mosaic-provider-jobcards")
    first_page = parse_search_page(first)
    all_results: list[dict[str, Any]] = list(first_page.get("results", []))
    total = max_results
    try:
        total = min(max_results, int(first_page["meta"][0]["jobCount"]))
    except (KeyError, IndexError, TypeError, ValueError):
        pass
    # Indeed paginates with start=10, 20, ...
    other_pages = []
    for start in range(10, total, 10):
        other_pages.append(_add_url_parameter(url, start=start))
    for page_url in other_pages:
        try:
            html = await _fetch_rendered_html(page_url, ready_selector="#mosaic-provider-jobcards")
            parsed = parse_search_page(html)
            all_results.extend(parsed.get("results", []))
            if len(all_results) >= max_results:
                break
        except Exception as e:  # noqa: BLE001
            logger.warning("search page {} failed: {}", page_url, e)
            break
    return all_results[:max_results]

async def scrape_jobs(job_keys: list[str]) -> list[dict[str, Any]]:
    """Scrape individual job listing pages by `jk` (job key)."""
    out: list[dict[str, Any]] = []
    for jk in job_keys:
        url = f"https://www.indeed.com/viewjob?jk={jk}"
        try:
            html = await _fetch_rendered_html(url, ready_selector="#jobDescriptionText")
            parsed = parse_job_page(html)
            if parsed:
                out.append(parsed)
        except Exception as e:  # noqa: BLE001
            # Indeed often locks /viewjob behind a bot-detection sign-in wall on
            # fresh proxies. Skip rather than poison the result file.
            logger.warning("skip jk={}: {}", jk, e)
    return out

def to_dict(obj: Any) -> Any:
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    return obj

File: indeed-scraper/browser/nodejs/indeed.mjs
// Indeed scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
//
// Indeed embeds two JSON blobs in every page:
//   - Search: window.mosaic.providerData["mosaic-provider-jobcards"]
//   - Job:    _initialData={...};
// Both are extracted with regex and surfaced with unchanged top-level keys.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";

const DEFAULT_PROXY_COUNTRY = "US";
const DEFAULT_SESSION_TTL = 180;

const SEARCH_JSON_RE = /window\.mosaic\.providerData\["mosaic-provider-jobcards"\]=(\{.+?\});/s;
const JOB_JSON_RE = /_initialData=(\{.+?\});/s;

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

function looksAuthWalled(html, finalUrl) {
  if (!html) return true;
  if (typeof finalUrl === "string" && /secure\.indeed\.com\/auth/.test(finalUrl)) return true;
  if (/Sign In \| Indeed Accounts/i.test(html)) return true;
  if (/Just a moment/i.test(html) && html.length < 90000) return true;
  return false;
}

async function fetchRenderedHtml(url, readySelector, { proxyCountry = DEFAULT_PROXY_COUNTRY, retries = 2, warmup = true } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const { browserWSEndpoint } = await client().browser.create({
      proxyCountry,
      sessionTTL: DEFAULT_SESSION_TTL,
    });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint });
      const page = await browser.newPage();
      // Warmup: hit indeed.com first so Cloudflare cookies attach to this
      // session before we open the deeper URL. Without this, /viewjob bounces
      // straight to secure.indeed.com/auth?from=bot-detection-anonymous.
      if (warmup) {
        try {
          await page.goto("https://www.indeed.com/", { waitUntil: "domcontentloaded", timeout: 30000 });
          // Let Cloudflare's challenge complete if there is one.
          for (let i = 0; i < 6; i++) {
            await new Promise((r) => setTimeout(r, 2000));
            const h = await page.content();
            if (h && !h.includes("Just a moment") && h.length > 30000) break;
          }
        } catch (_) {}
      }
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 45000 });
      // Wait for either the ready selector or the Cloudflare challenge to clear.
      let html = "";
      for (let i = 0; i < 6; i++) {
        if (readySelector) {
          try { await page.waitForSelector(readySelector, { timeout: 5000 }); break; } catch (_) {}
        }
        await new Promise((r) => setTimeout(r, 2000));
        html = await page.content();
        if (html && !html.includes("Just a moment")) break;
      }
      const finalUrl = await page.url();
      if (!html) html = await page.content();
      if (html && !looksAuthWalled(html, finalUrl)) return html;
      lastError = new Error(looksAuthWalled(html, finalUrl) ? `auth-wall (final=${finalUrl.slice(0, 60)})` : "empty HTML");
    } catch (e) {
      lastError = e;
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
  }
  throw new Error(`giving up after ${retries + 1} attempts: ${lastError?.message}`);
}

function addUrlParameter(url, params) {
  const u = new URL(url);
  for (const [k, v] of Object.entries(params)) u.searchParams.set(k, String(v));
  return u.toString();
}

// ---------------- parsers (verbatim shape) ----------------

export function parseSearchPage(html) {
  const m = html.match(SEARCH_JSON_RE);
  if (!m) return { results: [], meta: [] };
  const data = JSON.parse(m[1]);
  const model = data?.metaData?.mosaicProviderJobCardsModel ?? {};
  return {
    results: model.results ?? [],
    meta: model.tierSummaries ?? [],
  };
}

export function parseJobPage(html) {
  // Indeed serves the job blob via several regex shapes — try each in turn.
  let raw = null;
  for (const re of [
    JOB_JSON_RE,
    /window\._initialData\s*=\s*(\{[\s\S]+?\});/,
    /_initialData\s*=\s*JSON\.parse\("(.+?)"\);/,
  ]) {
    const m = html.match(re);
    if (m) { raw = m[1]; break; }
  }
  if (!raw) return {};
  let data;
  try {
    // Handle the JSON.parse-escaped variant.
    if (raw.startsWith("{")) data = JSON.parse(raw);
    else data = JSON.parse(JSON.parse('"' + raw + '"'));
  } catch (_) {
    return {};
  }
  const job = {};
  try {
    job.description = data.jobInfoWrapperModel.jobInfoModel.sanitizedJobDescription;
  } catch (_) {
    job.description = "";
  }
  for (const k of ["jobMetadataHeaderModel", "jobTagModel", "jobInfoHeaderModel"]) {
    const sub = data?.[k];
    if (sub && typeof sub === "object") Object.assign(job, sub);
  }
  return job;
}

// ---------------- scrape functions (mirror the upstream reference's exports) ----------------

export async function scrapeSearch(url, maxResults = 1000) {
  const first = await fetchRenderedHtml(url, "#mosaic-provider-jobcards");
  const firstPage = parseSearchPage(first);
  const all = [...(firstPage.results ?? [])];
  let total = maxResults;
  try {
    total = Math.min(maxResults, Number(firstPage.meta?.[0]?.jobCount));
  } catch (_) {}
  if (!Number.isFinite(total)) total = maxResults;
  const others = [];
  for (let start = 10; start < total; start += 10) {
    others.push(addUrlParameter(url, { start }));
  }
  for (const pageUrl of others) {
    try {
      const html = await fetchRenderedHtml(pageUrl, "#mosaic-provider-jobcards");
      const parsed = parseSearchPage(html);
      all.push(...(parsed.results ?? []));
      if (all.length >= maxResults) break;
    } catch (_) {
      break;
    }
  }
  return all.slice(0, maxResults);
}

export async function scrapeJobs(jobKeys) {
  const out = [];
  for (const jk of jobKeys) {
    const url = `https://www.indeed.com/viewjob?jk=${jk}`;
    try {
      const html = await fetchRenderedHtml(url, "#jobDescriptionText");
      const parsed = parseJobPage(html);
      if (parsed && Object.keys(parsed).length) out.push(parsed);
    } catch (e) {
      // Indeed often locks /viewjob behind a bot-detection sign-in wall when
      // the proxy is fresh. Skip rather than poison the result file.
      console.warn(`skip jk=${jk}: ${e.message}`);
    }
  }
  return out;
}
